### PR TITLE
koin-compose: use rememberUpdatedState with ParametersDefinition, optimized KoinApplication, KoinContext, KoinIsolatedContext

### PR DESCRIPTION
**Pull request by hoc081098**
- use rememberUpdatedState with ParametersDefinition
- delegate rememberKoinInject to koinInject, update deprecated message (will close https://github.com/InsertKoinIO/koin/issues/1743)
- optimized KoinApplication, KoinContext, KoinIsolatedContext
- remove unused UnknownKoinContext
- apiDump

### CODE REVIEW

**Comment by arnaudgiuliani** [comment]
is it optimizing around `rememberUpdatedState`? What is the target of this PR?

File: compose/koin-compose/src/commonMain/kotlin/org/koin/compose/KoinApplication.kt [hunk 1]
```diff
@@ -61,9 +61,8 @@ private fun getDefaultKoinContext() = KoinPlatformTools.defaultContext().get()
  * @author @author jjkester
  */
 @Composable
-fun getKoin(): Koin = currentComposer.run {
-    return LocalKoinApplication.current
-}
```
Comment (hoc081098) [inline compose/koin-compose/src/commonMain/kotlin/org/koin/compose/KoinApplication.kt:66]: currentComposer.run is unnecessary

File: compose/koin-compose/src/commonMain/kotlin/org/koin/compose/KoinApplication.kt [hunk 2]
```diff
@@ -72,9 +71,8 @@ fun getKoin(): Koin = currentComposer.run {
  *
  */
 @Composable
-fun currentKoinScope(): Scope = currentComposer.run {
-    return LocalKoinScope.current
-}
```
Comment (hoc081098) [inline compose/koin-compose/src/commonMain/kotlin/org/koin/compose/KoinApplication.kt:77]: currentComposer.run is unnecessary

File: compose/koin-compose/src/commonMain/kotlin/org/koin/compose/KoinApplication.kt [hunk 3]
```diff
@@ -92,7 +90,8 @@ fun rememberCurrentKoinScope(): Scope = currentComposer.run {
 
 @OptIn(KoinInternalApi::class)
 private fun Koin.warnNoContext() {
-    logger.info("[Warning] - No Koin context defined in Compose, fallback to default Koin context.\nUse KoinContext(), KoinAndroidContext() or KoinApplication() to setup or create Koin context with Compose and avoid such message.")
+    logger.info("[Warning] - No Koin context defined in Compose, fallback to default Koin context." +
+        "Use KoinContext(), KoinAndroidContext() or KoinApplication() to setup or create Koin context with Compose and avoid such message.")
 }
```
Comment (hoc081098) [inline compose/koin-compose/src/commonMain/kotlin/org/koin/compose/KoinApplication.kt:95]: Format string to 2 lines to improve the readability

File: compose/koin-compose/src/commonMain/kotlin/org/koin/compose/Inject.kt [hunk 4]
```diff
@@ -51,14 +54,12 @@ inline fun <reified T> koinInject(
...
     qualifier: Qualifier? = null,
     scope: Scope = rememberCurrentKoinScope(),
     noinline parameters: ParametersDefinition? = null,
-): T {
-    val st = parameters?.let { rememberStableParametersDefinition(parameters) }
-    return remember(qualifier, scope) {
-        scope.get(qualifier, st?.parametersDefinition)
-    }
-}
+): T = koinInject(qualifier, scope, parameters)
```
Comment (hoc081098) [inline compose/koin-compose/src/commonMain/kotlin/org/koin/compose/Inject.kt:65]: delegate to koinInject

File: compose/koin-compose/src/commonMain/kotlin/org/koin/compose/KoinApplication.kt [hunk 5]
```diff
@@ -121,10 +121,9 @@ fun KoinApplication(
     }
     CompositionLocalProvider(
         LocalKoinApplication provides koinApplication.koin,
-        LocalKoinScope provides koinApplication.koin.scopeRegistry.rootScope
-    ) {
-        content()
-    }
+        LocalKoinScope provides koinApplication.koin.scopeRegistry.rootScope,
+        content = content
+    )
```
Comment (hoc081098) [inline compose/koin-compose/src/commonMain/kotlin/org/koin/compose/KoinApplication.kt:126]: reduce an unnecessary lambda allocation
` { content() }` -> `content=content`

File: compose/koin-compose/src/commonMain/kotlin/org/koin/compose/KoinApplication.kt (hunk 1 above)
Comment (arnaudgiuliani) [reply compose/koin-compose/src/commonMain/kotlin/org/koin/compose/KoinApplication.kt:66]: yeah 🤔 I would need to benchmark this, being sure we don't miss something behind the scene

Comment (hoc081098) [reply compose/koin-compose/src/commonMain/kotlin/org/koin/compose/KoinApplication.kt:66]: @arnaudgiuliani 
From AndroidX:

https://github.com/androidx/androidx/blob/6d1f16a70c802850a83daae8929ea1662f972715/compose/runtime/runtime/src/commonMain/kotlin/androidx/compose/runtime/CompositionLocal.kt#L69-L75

```kotlin
@Stable
sealed class CompositionLocal<T> constructor(defaultFactory: () -> T) {
    internal val defaultValueHolder = LazyValueHolder(defaultFactory)

    internal abstract fun updatedStateOf(value: T, previous: State<T>?): State<T>

    /**
     * Return the value provided by the nearest [CompositionLocalProvider] component that invokes, directly or
     * indirectly, the composable function that uses this property.
     *
     * @sample androidx.compose.runtime.samples.consumeCompositionLocal
     */
    @OptIn(InternalComposeApi::class)
    inline val current: T
        @ReadOnlyComposable
        @Composable
        get() = currentComposer.consume(this)
}
```

Comment (hoc081098) [reply compose/koin-compose/src/commonMain/kotlin/org/koin/compose/KoinApplication.kt:66]: add `@ReadOnlyComposable` to `getKoin()` will optimize the generated code.
Docs: https://developer.android.com/reference/kotlin/androidx/compose/runtime/ReadOnlyComposable

Comment (arnaudgiuliani) [reply compose/koin-compose/src/commonMain/kotlin/org/koin/compose/KoinApplication.kt:66]: thanks

Comment (hoc081098) [reply compose/koin-compose/src/commonMain/kotlin/org/koin/compose/KoinApplication.kt]: I have just fixed the conflict. Could you check this PR again 🙏 ?

**Comment by arnaudgiuliani** [comment]
Good work. I need to take time to take a tour :)

**Review by arnaudgiuliani — APPROVED** [review]

**Comment by arnaudgiuliani** [comment]
It's just a api diff confilct I believe @hoc081098 👍

**Comment by hoc081098** [comment]
> It's just a api diff confilct I believe @hoc081098 👍 

I will update it 🙏

**Comment by arnaudgiuliani** [comment]
done edit for you @hoc081098 👍

**Comment by hoc081098** [comment]
> done edit for you @hoc081098 👍 

Thanks for your help

**Pull request merged**